src/Sylius/Behat/Context/Api/Admin/ManagingExchangeRatesContext.php

```diff
@@ -0,0 +1,363 @@
+     */
+    public function iShouldNotBeAbleToEditItsTargetCurrency(): void
+    {
+        $this->assertIfNotBeAbleToEditItCurrency('targetCurrency');
+    }
+
+    /**
+     * @Then I should be notified that :element is required
+     */
+    public function iShouldBeNotifiedThatIsRequired(string $element): void
+    {
+        Assert::contains($this->client->getError(), sprintf('%s: Please enter exchange rate %s.', $element, $element));
+    }
+
+    /**
+     * @Then I should be notified that the ratio must be greater than zero
+     */
+    public function iShouldBeNotifiedThatRatioMustBeGreaterThanZero(): void
+    {
+        Assert::contains($this->client->getError(), 'The ratio must be greater than 0.');
+    }
+
+    /**
+     * @Then I should be notified that source and target currencies must differ
+     */
+    public function iShouldBeNotifiedThatSourceAndTargetCurrenciesMustDiffer(): void
+    {
+        Assert::contains($this->client->getError(), 'The source and target currencies must differ.');
+    }
+
+    /**
+     * @Then I should be notified that the currency pair must be unique
+     */
+    public function iShouldBeNotifiedThatTheCurrencyPairMustBeUnique(): void
+    {
+        Assert::contains($this->client->getError(), 'The currency pair must be unique.');
+    }
+
+    private function assertIfNotBeAbleToEditItCurrency(string $currencyType): void
+    {
+        $this->client->buildUpdateRequest('exchange_rates', $this->sharedStorage->get('exchange_rate_id'));
+
+        $this->client->addRequestData($currencyType, '/new-api/currencies/EUR');
+        $this->client->update();
+
+        $this->client->index('exchange_rates');
+        Assert::false(
+            $this->client->hasItemOnPositionWithValue(0, $currencyType, '/new-api/currencies/EUR'),
+            sprintf('The %s field with value \'/new-api/currencies/EUR\' exists', $currencyType)
+        );
+    }
+
+    private function getExchangeRateBetweenCurrencies(string $sourceCode, string $targetCode): ExchangeRateInterface
+    {
+        /** @var ExchangeRateInterface|null */
+        $exchangeRate = $this->exchangeRateRepository->findOneWithCurrencyPair($sourceCode, $targetCode);
+
+        Assert::notNull($exchangeRate);
+
+        return $exchangeRate;
+    }
+
+    private function getExchangeRateFromResponse(
+        CurrencyInterface $sourceCurrency,
+        CurrencyInterface $targetCurrency
+    ): ?array {
+        $this->client->index('exchange_rates');
+
+        /** @var array $item */
+        foreach ($this->client->getCollection() as $item)
+        {
+            if (
+                $item['sourceCurrency'] === '/new-api/currencies/' . $sourceCurrency->getCode() &&
+                $item['targetCurrency'] === '/new-api/currencies/' . $targetCurrency->getCode()
+            ) {
+                return $item;
+            }
+        }
+
+        return null;
+    }
+
+    private function hasExchangeRateFromResponse(
```

Comment:
There is a lot of logic in this context. Maybe we could create another ApiClient just for exchange rates? Or create some service that would contain that logic? I just feel it can quickly go out of control 😄

Reply:
I reduced this step :)